src/mappings/poktroll/applications.ts

```diff
@@ -0,0 +1,611 @@
+      messageId: msgId,
+    }).save(),
+  ])
+}
+
+async function _handleTransferApplicationBeginEvent(
+  event: CosmosEvent
+) {
+  logger.debug(`[handleTransferApplicationBeginEvent] (event.msg): ${stringify(event.msg, undefined, 2)}`);
+  const msg: CosmosMessage<EventTransferBegin> = event.msg
+
+  const transferEndHeight = event.event.attributes.find(attribute => attribute.key === "transfer_end_height")?.value as string
+
+  if (!transferEndHeight) {
+    throw new Error(`[handleTransferApplicationBeginEvent] transferEndHeight not found`);
+  }
+
+  const application = await Application.get(msg.msg.decodedMsg.sourceAddress)
+
+  if (!application) {
+    throw new Error(`[handleTransferApplicationBeginEvent] application not found`);
+  }
+
+  application.transferEndHeight = BigInt((transferEndHeight as unknown as string).replaceAll('"', ""))
+
+  const eventId = getEventId(event)
+
+  await Promise.all([
+    application.save(),
+    EventTransferBeginEntity.create({
+      id: eventId,
+      sourceId: msg.msg.decodedMsg.sourceAddress,
+      destinationId: msg.msg.decodedMsg.destinationAddress,
+      transactionId: event.tx.hash,
+      blockId: event.block.block.id,
+      eventId,
+    }).save()
+  ])
+}
+
+async function _handleTransferApplicationEndEvent(
+  event: CosmosEvent
+) {
+  logger.debug(`[handleTransferApplicationEndEvent] (event.event): ${stringify(event.event, undefined, 2)}`);
+
+  let sourceAddress = event.event.attributes.find(attribute => attribute.key === "source_address")?.value as unknown as string
+
+  if (!sourceAddress) {
+    throw new Error(`[handleTransferApplicationEndEvent] event.event.attributes not found`);
+  }
+
+  // the source address is surrounded by quotes
+  sourceAddress = sourceAddress.replaceAll('"', "")
+
+  const sourceApplication = await Application.get(sourceAddress)
+
+  if (!sourceApplication) {
+    throw new Error(`[handleTransferApplicationMsg] source application not found for address ${sourceAddress}`)
+  }
+
+  sourceApplication.transferringToId = undefined
+  sourceApplication.transferEndHeight = undefined
+  sourceApplication.transferEndBlockId = event.block.block.id
+  sourceApplication.status = StakeStatus.Unstaked
+  sourceApplication.unstakingReason = 2
+  sourceApplication.unstakingEndBlockId = event.block.block.id
+
+  const destinationAppStringified = event.event.attributes.find(attribute => attribute.key === "destination_application")?.value as string
+
+  if (!destinationAppStringified) {
+    throw new Error(`[handleTransferApplicationMsg] destination application not in event`);
+  }
+
+  const destinationApp: Required<ApplicationSDKType> = JSON.parse(destinationAppStringified)
+
+  sourceApplication.destinationApplicationId = destinationApp.address
+
+  const {delegatee_gateway_addresses, service_configs,} = destinationApp
+  const stake = destinationApp.stake as Required<typeof destinationApp.stake>
+
+  const destinationApplication  = Application.create({
```

Comment:
Similar feedback as suppliers, don't we need to check if there's an existing app that's staked or unstaked; in which case, it would be upstaking or restaking, respectively?